src/System.Management.Automation/engine/Subsystem/Prediction/IPredictor.cs

```diff
@@ -0,0 +1,163 @@
+        /// <returns>A list of predictive suggestions.</returns>
+        List<PredictiveSuggestion>? GetSuggestion(PredictionContext context, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Context information about the user input.
+    /// </summary>
+    public class PredictionContext
+    {
+        /// <summary>
+        /// Gets the abstract syntax tree (AST) generated from parsing the user input.
+        /// </summary>
+        public Ast InputAst { get; }
+
+        /// <summary>
+        /// Gets the tokens generated from parsing the user input.
+        /// </summary>
+        public IReadOnlyList<Token> InputTokens { get; }
+
+        /// <summary>
+        /// Gets the cursor position, which is assumed always at the end of the input line.
+        /// </summary>
+        public IScriptPosition CursorPosition { get; }
+
+        /// <summary>
+        /// Gets the token at the cursor.
+        /// </summary>
+        public Token? TokenAtCursor { get; }
+
+        /// <summary>
+        /// Gets all ASTs that are related to the cursor position,
+        /// which is assumed always at the end of the input line.
+        /// </summary>
+        public IReadOnlyList<Ast> RelatedAsts { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PredictionContext"/> class from the AST and tokens that represent the user input.
+        /// </summary>
+        /// <param name="inputAst">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="inputTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        public PredictionContext(Ast inputAst, Token[] inputTokens)
+        {
+            Requires.NotNull(inputAst, nameof(inputAst));
+            Requires.NotNull(inputTokens, nameof(inputTokens));
+
+            var cursor = inputAst.Extent.EndScriptPosition;
+            var astContext = CompletionAnalysis.ExtractAstContext(inputAst, inputTokens, cursor);
+
+            InputAst = inputAst;
+            InputTokens = inputTokens;
+            CursorPosition = cursor;
+            TokenAtCursor = astContext.TokenAtCursor;
+            RelatedAsts = astContext.RelatedAsts;
+        }
+
+        /// <summary>
+        /// Creates a context instance from the user input line.
+        /// </summary>
+        /// <param name="input">The user input.</param>
+        /// <returns>A <see cref="PredictionContext"/> object.</returns>
+        public static PredictionContext Create(string input)
+        {
+            Requires.NotNullOrEmpty(input, nameof(input));
+
+            Ast ast = Parser.ParseInput(input, out Token[] tokens, out _);
+            return new PredictionContext(ast, tokens);
+        }
+    }
+
+    /// <summary>
+    /// The class represents a predictive suggestion generated by a predictor.
+    /// </summary>
+    public sealed class PredictiveSuggestion
```

Comment:
It could be. Then it will result in copies of a lot of such struct instances, from `IPredictor.GetSuggestion` to `CommandPrediction.PredictInput`, and then from `CommandPrediction.PredictInput` to the caller. I don't know if that's better or worse, and thus prefer to not do it.

Reply:
My previous comment about "resulting in a lot copies" is not accurate. Copies only happen when working with individual `PredictiveSuggestion` instances if making the type a struct. However, the list holds those `PredictiveSuggestion` elements in an array (`PredictiveSuggestion[]`), which will always be allocated on heap, so even if we make this type a struct, it will still take heap allocation, the same as when making the type a class. So my conclusion is the same -- it could be a premature optimization that we should avoid at this time.

Reply:
For your future investigations.
I don't know _how_ the API will be used by consumers. 
Thus, we could replace List with Enumerable if it is expected that consumes will retrieve suggestions by one and maybe use only first. In the case async task can return first result with less delay. Also perhaps a value task return cache will work.
Another scenario, consumer will filter suggestion results. In the case we could implement an enumerator like .Net has for paths ([FileSystemEnumerable](https://source.dot.net/#System.IO.FileSystem/System/IO/Enumeration/FileSystemEnumerable.cs,97f8891cf1b2a468)) and benefit from delegates. 
Injecting a consumer code with delegates can have some benefits and reduce delays.